grandcypher/__init__.py

```diff
@@ -414,21 +438,76 @@ def skip_clause(self, skip):
         self._skip = skip
 
     def returns(self, ignore_limit=False):
-        if self._limit and ignore_limit is False:
-            offset_limit = slice(self._skip, self._skip + self._limit)
-        else:
-            offset_limit = slice(self._skip, None)
-
-        results = self._lookup(self._return_requests, offset_limit=offset_limit)
 
+        results = self._lookup(
+            self._return_requests + list(self._order_by_attributes), 
+            offset_limit=slice(0, None)
+        )
+        if self._order_by:
+            results = self._apply_order_by(results)
         if self._distinct:
+            results = self._apply_distinct(results)
+        results = self._apply_pagination(results, ignore_limit)
+
+
+        # Exclude order-by-only attributes from the final results
+        results = {
+            key: values for key, values in results.items() if key in self._return_requests
+        }
 
-            # process distinct for each key in results
-            distinct_results = {}
-            for key, values in results.items():
-                # remove duplicates
-                distinct_results[key] = list(set(values))
-            results =  distinct_results
+        return results
+    
+    def _apply_order_by(self, results):
+        if self._order_by:
+            sort_lists = [(results[field], direction) for field, direction in self._order_by if field in results]
+
+            if sort_lists:
+                # Generate a list of indices sorted by the specified fields
+                indices = range(len(next(iter(results.values()))))  # Safe because all lists are assumed to be of the same length
+                for sort_list, direction in reversed(sort_lists):  # reverse to ensure the first sort key is primary
+                    indices = sorted(indices, key=lambda i: sort_list[i], reverse=(direction == 'DESC'))
+
+                # Reorder all lists in results using sorted indices
+                for key in results:
+                    results[key] = [results[key][i] for i in indices]
+        
+        return results
+    
+    def _apply_distinct(self, results):
+        if self._order_by:
+            assert self._order_by_attributes.issubset(self._return_requests), "In a WITH/RETURN with DISTINCT or an aggregation, it is not possible to access variables declared before the WITH/RETURN"
```

Comment:
These interpretable error messages are great — inexplicable errors are one of my main Sads with this repo right now! Really appreciate you :)

Reply:
i was banging my head on this for wayy too long before realizing that Cypher doesn't even support this😅 Well, it does, but not like this

Reply:
Ha, I had a few of those in the early days of the library. I totally know that feeling! 😆